docs/src/best-practices-js.md

```diff
@@ -0,0 +1,201 @@
+---
+id: best-practices
+title: "Best Practices"
+---
+
+
+This guide should help you to make sure you are following our best practices and writing tests that are more resilient.
+
+## Testing philosophy
+
+### Test user-visible behavior
+
+Automated tests should verify that the application code works for the end users, and avoid relying on implementation details. Things which users will not typically use, see, or even know about such as the name of a function, whether something is an array, or the CSS class of some element. The end user will see or interact with what is rendered on the page, so your test should typically only see/interact with the same rendered output.
+
+### Make tests as isolated as possible
+
+Each test should be completely isolated from another test and should run independently with it's own local storage, session storage, cookies etc. [Test isolation](./browser-contexts.md) improves reproducibility, makes debugging easier and prevents cascading test failures.
+
+In order to avoid repetition for a particular part of your test you can use [before and after hooks](/api/class-test.md). Within your test file add a before hook to run a part of your test before each test such as going to a particular URL or logging in to a part of your app. This keeps your tests isolated as no test relies on another. However it is also ok to have a little duplication when tests are simple enough especially if it keeps your tests clearer and easier to read and maintain.
+
+### Make your tests fail
+
+Visually seeing your test pass is a great way to gain confidence in your tests. But sometimes you can't follow everything as your tests run too fast or perhaps you are running in headless mode and therefore not visually seeing anything. Making your tests fail ensures you understand what the test is not expecting. The error message should tell you what it received and what it expected. It's very easy to write a test that never fails which isn't much use at all.
+
+### Write less tests but longer tests
+
+When it comes to end to end testing having long tests is not a bad thing. It's ok to have multiple actions and assertions in your test. You should avoid separating your assertions into individual test blocks as it doesn't really bring much value and just slows down the running of your tests. If your test does fail Playwright will give you an error message showing what part of the test failed which you can see either in VS Code, the terminal, the HTML report or the trace viewer.
```

Comment:
Suggest using `expect.soft()` when possible, to gather as much failures from a single test as possible.

Reply:
We've come to the same conclusion making the test a bit longer to test different scenarios while on a page (invalid input, with valid input, no input for example) to check the ui validation for example. We found utilizing the test.step() for these helped break up the different tests and made the playwright test report much nicer to view.

Reply:
thanks. will look into it some more and follow up in another pr